core/base/block_operator.cpp

```diff
@@ -0,0 +1,356 @@
+    }
+    return {num_rows, num_cols};
+}
+
+
+}  // namespace
+
+
+std::unique_ptr<BlockOperator> BlockOperator::create(
+    std::shared_ptr<const Executor> exec)
+{
+    return std::unique_ptr<BlockOperator>(new BlockOperator(std::move(exec)));
+}
+
+
+std::unique_ptr<BlockOperator> BlockOperator::create(
+    std::shared_ptr<const Executor> exec,
+    std::vector<std::vector<std::shared_ptr<LinOp>>> blocks)
+{
+    return std::unique_ptr<BlockOperator>(
+        new BlockOperator(std::move(exec), std::move(blocks)));
+}
+
+
+BlockOperator::BlockOperator(std::shared_ptr<const Executor> exec)
+    : EnableLinOp<BlockOperator>(std::move(exec))
+{}
+
+
+BlockOperator::BlockOperator(
+    std::shared_ptr<const Executor> exec,
+    std::vector<std::vector<std::shared_ptr<LinOp>>> blocks)
+    : EnableLinOp<BlockOperator>(exec, compute_global_size(blocks)),
+      block_size_(blocks.empty()
+                      ? dim<2>{}
+                      : dim<2>(blocks.size(), blocks.front().size())),
+      row_spans_(compute_local_spans(
+          block_size_[0], blocks,
+          [&](auto i) {
+              return find_non_zero_in_row(blocks, i)->get_size()[0];
+          })),
+      col_spans_(compute_local_spans(block_size_[1], blocks, [&](auto i) {
+          return find_non_zero_in_col(blocks, i)->get_size()[1];
+      }))
+{
+    for (auto& row : blocks) {
+        for (auto& block : row) {
+            if (block && block->get_executor() != exec) {
+                blocks_.push_back(gko::clone(exec, block));
+            } else {
+                blocks_.push_back(std::move(block));
+            }
+        }
+    }
+}
+
+
+void validate_application_block_parameters(ptr_param<const BlockOperator> op,
+                                           ptr_param<const BlockOperator> b,
+                                           ptr_param<const BlockOperator> x)
+{
+    dim<2> single_col{0, 1};
+    GKO_ASSERT_CONFORMANT(op->get_block_size(), b->get_block_size());
+    GKO_ASSERT_EQUAL_ROWS(op->get_block_size(), x->get_block_size());
+    GKO_ASSERT_EQUAL_COLS(single_col, b->get_block_size());
+    GKO_ASSERT_EQUAL_COLS(single_col, x->get_block_size());
+}
+
+
+void BlockOperator::apply_impl(const LinOp* b, LinOp* x) const
+{
+    auto block_b = block_vector(b, col_spans_);
+    auto block_x = block_vector(x, row_spans_);
+
+    validate_application_block_parameters(this, block_b, block_x);
+
+    auto one =
+        gko::initialize<matrix::Dense<double>>({1}, this->get_executor());
+    for (size_type row = 0; row < block_size_[0]; ++row) {
+        dispatch_dense([](auto* dense) { dense->fill(0.0); },
+                       block_x->block_at(row, 0));
+        for (size_type col = 0; col < block_size_[1]; ++col) {
+            if (!block_at(row, col)) {
+                continue;
+            }
+            block_at(row, col)->apply(one.get(), block_b->block_at(col, 0),
+                                      one.get(), block_x->block_at(row, 0));
```

Comment:
The issue is finding the first apply. Since the blocks can be nullptr, the first apply in a row (`col = 0`) maybe zero, so I can't use that as a check. I will see if I can find a somewhat elegant solution for that.